include/wil/registry_basic.h

```diff
@@ -0,0 +1,1675 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+// ReSharper disable CppInconsistentNaming
+#ifndef __WIL_REGISTRY_BASIC_INCLUDED
+#define __WIL_REGISTRY_BASIC_INCLUDED
+
+// TODO: should we use feature checks instead of forced imports?
+#include <functional>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include <list>
+
+#include <Windows.h>
+#include <sddl.h>
+
+#include "common.h"
+#include "resource.h"
+
+#ifdef _KERNEL_MODE
+#error This header is not supported in kernel-mode.
+#endif
+
+#ifndef __WIL_WINREG_
+#error This required wil::unique_hkey support
+#endif
+
+namespace wil
+{
+namespace registry

```

Comment:
I see 3 options for "namespacing" this:

1. (this approach) `wil::registry`
2. Abbreviated: `wil::reg`
3. Prefixed: `wil::reg_*`
4. None: `wil::*` (`wil::create_reg_key`, `wil::set_reg_dword` eg)

The current approach works well enough and is super clear, but it leads to code like this:

```cpp
const auto key = wil::registry::open_key(/* ... */);
const auto value = wil::registry::get_value_dword(key, /* ... */);
wil::registry::set_value_dword(key, /* ... */);
// etc
```

vs `wil::reg::get_dword` etc might be easier to type.

Reply:
I personally prefer typing out 'registry' as a namespace. but I do see your point. It looks like the registry watcher code doesn't use a namespace.

if I had to pick a no-namespace-version, I would probably go with:
wil::open_reg_key()
wil::get_reg_dword()

I think I prefer namespaces, though:
wil::reg::open_key()
wil::reg::get_dword()

I thought putting the COM stuff in its own namespace made things much better

Reply:
I like `wil::reg::*` as you've done. I'm happy to stick with that :)